### PR TITLE
Implement tensor.align_to(names), torch.align_tensors(t1, t2, ...)

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24306 Implement name inference for mm, addmm
* #24203 Implement name inference for t(), transpose(...)
* #24087 Name inference for softmax, log_softmax and Dimname overloads.
* #24028 Fix out= function semantics for named tensors.
* #23973 Update tensor.view_names / tensor.names_ API
* #23962 Rename set_names -> view_names, set_names_ -> names_
* **#23804 Implement tensor.align_to(names), torch.align_tensors(t1, t2, ...)**
* #24109 Named inference for contiguous(), bernoulli variants, and dropout.
* #24108 Implement name inference rule for empty_like, clone
* #24107 Add `names` argument to ones, rand, randn, zeros, full; fix empty

`output = tensor.align_to(names)` returns a view of `tensor` such that
`output.names = names`. Dimensions with the same names in `tensor` and
`output` have the same sizes; dimensions with new names have size 1.

The following must be true for this operation to succeed:
1) tensor.names must be a subsequence (not necessarily contiguous) of `names`
2) Aligning tensor.names to names must not change the absolute position from the
   right of any unnamed dimension.

In practice, these constraints mean that aligning cannot transpose
names.

Some examples:
- Tensor[C].align_to(C) -> Tensor[C]
- Tensor[N].align_to([N, C]) -> Tensor[N, C]
- Tensor[H, W].align_to([N, H, W, C]) -> Tensor[N, H, W, C]
- Tensor[None].align_to([N, None]) -> Tensor[N, None]
- Tensor[N].align_to([N, None None]) -> Tensor[N, None, None]

Examples of error cases:
- Tensor[W, H].align_to([N, H, W, C]) -> Error (not a subsequence)
- Tensor[None, H].align_to([None, H, W]) -> Error (would change the
absolute position from the right of a None dimension)

`torch.align_tensors(*tensors)` aligns the named dimensions of each
tensor according to the alignment rules so that they can be used in an
operation. More concretely, it aligns each tensor to the
longest names among the names of the tensors in `tensors`.

This allows users to emulate "broadcasting by names", which is one of
the things named tensors tries to enable. Here is an example:

```
imgs: Tensor[N, C, H, W]
scale: Tensor[N]

// Doesn't work because we do broadcasting by alignment by default
imgs * scale

// Does work
imgs, scale = torch.align_tensors(imgs, scale)
imas * scale
```

Future:
- Consider allowing broadcasting by names by default.

Test Plan:
- The diff looks pretty large but more than half of it is testing.
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 23804 gh/zou3519/91/head

Differential Revision: [D16657927](https://our.internmc.facebook.com/intern/diff/D16657927)

### CODE REVIEW

File: aten/src/ATen/native/NamedTensor.cpp [hunk 1]
```diff
@@ -16,5 +16,144 @@ Tensor set_names(const Tensor& self, optional<DimnameList> names) {
   return result;
 }
 
+static void report_moving_unnamed_dim_error(
+    DimnameList names, DimnameList other, bool is_aligning_two_tensors) {
+  if (is_aligning_two_tensors) {
+    TORCH_CHECK(false,
+        "Could not infer the alignment of Tensor", names, " and Tensor", other,
+        ", please name the unnamed dimensions to resolve ambiguity.");
+    TORCH_CHECK(false,
```
Comment (nairbv) [inline aten/src/ATen/native/NamedTensor.cpp]: is this reachable?

File: aten/src/ATen/native/NamedTensor.cpp [hunk 2]
```diff
@@ -16,5 +16,144 @@ Tensor set_names(const Tensor& self, optional<DimnameList> names) {
...
+        "Aligning tensor `a` with dims ", names, " to tensor `b` with dims ", other,
```
Comment (nairbv) [inline aten/src/ATen/native/NamedTensor.cpp]: i think we could exclude the `a` and `b`, could add confusion since those probably aren't the actual tensor names and I think the message works without them.

File: aten/src/ATen/native/NamedTensor.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/NamedTensor.cpp]: Oops I think this was a copy-paste fiasco

File: aten/src/ATen/native/NamedTensor.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/native/NamedTensor.cpp]: Yeah, that sounds better

**Comment by zou3519** [comment]
@nairbv comments addressed

File: aten/src/ATen/native/NamedTensor.cpp [hunk 3]
```diff
@@ -16,5 +16,141 @@ Tensor set_names(const Tensor& self, optional<DimnameList> names) {
...
+static std::vector<Tensor> align_tensors_to(TensorList tensors, DimnameList names) {
+  std::vector<Tensor> result;
+  result.reserve(tensors.size());
+  for (const auto& tensor : tensors) {
+    result.emplace_back(align(tensor, names, /*is_aligning_two_tensors=*/true));
+  }
+  return result;
+}
+
+std::vector<Tensor> align_tensors(TensorList tensors) {
```
Comment (nairbv) [inline aten/src/ATen/native/NamedTensor.cpp]: can `TensorList` be any number of tensors? If so should `..._two_tensors` be named something else?

File: torch/functional.py [hunk 4]
```diff
@@ -825,3 +826,10 @@ def gels(input, A, out=None):
     warnings.warn("torch.gels is deprecated in favour of torch.lstsq and will be removed in "
                   "the next release. Please use torch.lstsq instead.", stacklevel=2)
     return torch.lstsq(input, A, out=out)
+
+
+def align_tensors(*tensors):
+    if not torch._C._BUILD_NAMEDTENSOR:
+        raise RuntimeError('NYI: torch.align_tensors is experimental and a part '
+                           'of our named tensors project.')
```
Comment (nairbv) [inline torch/functional.py]: NYI == 'not yet implemented,' but this is implemented right? I'd think the error should just tell them they need to have built with the experimental build option to use this

File: aten/src/ATen/native/NamedTensor.cpp (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/native/NamedTensor.cpp]: If the tensors are not alignable, the error message tells the user two of the tensors we can't align in the list. Because of that, I think it's fine that that variable is named `..._two_tensors`.

File: torch/functional.py (hunk 4 above)
Comment (zou3519) [reply torch/functional.py]: I don't want any users to start using this as an experimental feature right now so I wanted to tell them "hey, nothing to see here yet, move along".

File: aten/src/ATen/native/NamedTensor.cpp [hunk 5]
```diff
@@ -16,5 +16,141 @@ Tensor set_names(const Tensor& self, optional<DimnameList> names) {
...
+static void report_moving_unnamed_dim_error(
+    DimnameList names, DimnameList other, bool is_aligning_two_tensors) {
+  if (is_aligning_two_tensors) {
+    TORCH_CHECK(false,
+        "Aligning Tensor", names, " and Tensor", other,
+        " would change the absolute position from the right of an unnamed dimension. ",
+        "Please name unnamed dimensions to avoid ambiguity.");
+  } else {
+    TORCH_CHECK(false,
+        "Aligning Tensor", names, " to `names` ", other,
```
Comment (nairbv) [inline aten/src/ATen/native/NamedTensor.cpp]: I think you're missing a space car after 'Tensor'

File: aten/src/ATen/native/NamedTensor.cpp [hunk 6]
```diff
@@ -16,5 +16,141 @@ Tensor set_names(const Tensor& self, optional<DimnameList> names) {
...
+  }
+}
+
+static void report_not_a_subsequence_error(
+    DimnameList names, DimnameList other, bool is_aligning_two_tensors) {
+  if (is_aligning_two_tensors) {
+    auto shorter = names.size() > other.size() ? other : names;
+    auto longer = names.size() > other.size() ? names : other;
+    TORCH_CHECK(false,
+        "Could not align Tensor", shorter, " and Tensor", longer,
```
Comment (nairbv) [inline aten/src/ATen/native/NamedTensor.cpp]: same here ... or does the string representation of `shorter` start with a space?

File: aten/src/ATen/native/NamedTensor.cpp (hunk 5 above)
Comment (zou3519) [reply aten/src/ATen/native/NamedTensor.cpp]: This was intentional. The error message will look like
```
Aligning Tensor['N', 'C', 'H', 'W'] to Tensor['W', 'H'] ...
```

File: aten/src/ATen/native/NamedTensor.cpp [hunk 7]
```diff
@@ -16,5 +16,141 @@ Tensor set_names(const Tensor& self, optional<DimnameList> names) {
...
+// This helper function computes the resulting size of `t` after aligning it
+// to `aligned_names`. Enforces the alignment rules in Note [Alignment rules].
+static std::vector<int64_t> aligned_size(
+    IntArrayRef tensor_sizes,
+    DimnameList tensor_names,
+    DimnameList aligned_names,
+    bool is_aligning_two_tensors) {
+  std::vector<int64_t> expanded_sizes(aligned_names.size(), 1);
+  ptrdiff_t dim = (ptrdiff_t)tensor_sizes.size() - 1;
+  ptrdiff_t idx = (ptrdiff_t)aligned_names.size() - 1;
```
Comment (nairbv) [inline aten/src/ATen/native/NamedTensor.cpp]: If I understand this right, you're explicitly casting to a signed type so that you can check `>= 0` below? I guess that's fine, but seems a bit inelegant

File: aten/src/ATen/native/NamedTensor.cpp (hunk 5 above)
Comment (nairbv) [reply aten/src/ATen/native/NamedTensor.cpp]: ah OK

**Review by nairbv — APPROVED** [review]

File: aten/src/ATen/native/NamedTensor.cpp (hunk 7 above)
Comment (zou3519) [reply aten/src/ATen/native/NamedTensor.cpp]: Yes
